[MRG+1] Issue #2919: Fix FormRequest.formdata with GET method duplicates same key in query string

maramsumanth opened the pull request:
Solved https://github.com/scrapy/scrapy/issues/2919.

Combining ```querystr``` and ```formdata``` into dictionary, and thereby eliminating the duplicates. @kmike , @lopuhin , please review. 
I have also corrected the code, by creating a list where we want to pass ```duplicate keys```. as described in comment https://github.com/scrapy/scrapy/pull/3269#issuecomment-412260614

I am currently preparing for GSoC 2019, please guide me if there are any mistakes/suggestions. I would be glad if you give me suggestions for getting selected in scrapy.
Thanks in advance :) 👍

maramsumanth commented:
@lopuhin , can you please review this ? :)

maramsumanth commented:
@lopuhin , what can be the reason here for failure of codecov/patch ?

maramsumanth commented:
@elacuesta , could you review ?

maramsumanth commented:
@kmike , can you review this PR? :)

elacuesta commented:
I couldn't take a deep look into the algorithm itself, these are just a few Python suggestions to improve readibility:

1. There are some duplicated occurences of `urlsplit(self.url)`, you could assign that to a variable instead. `urlsplit`

2. It's better to iterate sequences directly instead of by index, i.e.
```python
In [1]: keys = ['a', 'b', 'c']                                                                                                                                                                                                                                                  

In [2]: for k in keys:  # good 
   ...:     print(k)                                                                                                                                                                                                                                                            
a
b
c

In [3]: for i in range(len(keys)):  # not so good 
   ...:     print(keys[i])                                                                                                                                                                                                                                                      
a
b
c
```

3. If you need to iterate at the same time over more than one iterator you could use [`zip`](https://docs.python.org/3.6/library/functions.html#zip)
```python
In [4]: values = [1, 2, 3]                                                                                                                                                                                                                                                      

In [5]: for k, v in zip(keys, values): 
   ...:     print('{}={}'.format(k, v))                                                                                                                                                                                                                                         
a=1
b=2
c=3
```

Finally, if I'm not mistaken the Codecov failure is because some of the new lines are not hit by any of the existing tests, that's why it says "will decrease coverage by 0.04%"

maramsumanth commented:
@elacuesta , thanks for the suggestion :) 👍. I have modified code to improve readability.

amarynets commented on scrapy/http/request/form.py:
```diff
@@ -33,7 +33,25 @@ def __init__(self, *args, **kwargs):
+                if url_split.query:
+                    queries = (url_split.query).split('&')
+                    query_dict = {}
+                    for x in queries:
```
Why are you not using `parse_qs` or `parse_qsl` for parsing and create `query_dict ` dict?

amarynets commented on scrapy/http/request/form.py:
```diff
@@ -33,7 +33,25 @@ def __init__(self, *args, **kwargs):
+                            query_dict[k] = v
+                        query_str = ''
+                        for k, v in query_dict.items():
+                            query_str += (k + '=' + v + '&')
```
Also you can use `urlencode` for creating qs from dict

amarynets commented on scrapy/http/request/form.py:
```diff
@@ -33,7 +33,25 @@ def __init__(self, *args, **kwargs):
+                        query_str = ''
+                        for k, v in query_dict.items():
+                            query_str += (k + '=' + v + '&')
+                    self._set_url(self.url[:self.url.index('?')+1] + query_str + querystr + ('#' + url_split.fragment if url_split.fragment else ''))
```
I think would be better to use `urllib.parse.urlunparse()` for create new URL instead concatenation.

maramsumanth commented:
Done @amarynets. Thanks for your suggestions! :) 👍

maramsumanth commented:
@kmike @lopuhin , can you please look into this?

maramsumanth commented:
@lopuhin @elacuesta , can you look into this?

maramsumanth commented:
@dangra , can you review? Thanks.

maramsumanth commented:
I have made the necessary changes, @kmike can you review it? :)

maramsumanth commented:
@lopuhin , can you please look into this?

lopuhin commented on scrapy/http/request/form.py:
```diff
@@ -33,7 +33,14 @@ def __init__(self, *args, **kwargs):
+                for k in querystr.split('&'):
+                    formdata_key_list.append(k.split('=')[0])
+                items = []
+                items += [(k, v) for k, v in parse_qsl(url_split.query) if k not in formdata_key_list]
```
Why create an empty ``items`` here and then add to this list? Can this be reduced to one operation?

maramsumanth replied:
Here ```items``` is a local variable, since this is the process of appending to a empty list. If we write simply ```items+=[........]``` it will be error as ```local variable referenced before assignment```

maramsumanth replied:
:+1:

lopuhin commented on scrapy/http/request/form.py:
```diff
@@ -33,7 +33,14 @@ def __init__(self, *args, **kwargs):
-                self._set_url(self.url + ('&' if '?' in self.url else '?') + querystr)
+                url_split = urlsplit(self.url)
+                formdata_key_list = []
+                for k in querystr.split('&'):
```
It's better to use functions for URL parsing from standard library and w3lib instead of doing this manually here, unless they are not doing what you need here (in that case it would make sense to leave a comment)

lopuhin commented on scrapy/http/request/form.py:
```diff
@@ -33,7 +33,14 @@ def __init__(self, *args, **kwargs):
+                items = []
+                items += [(k, v) for k, v in parse_qsl(url_split.query) if k not in formdata_key_list]
+                query_str = _urlencode(items, self.encoding)
+                self._set_url(urljoin(self.url,'?'+ (query_str + '&' if query_str else '') + querystr + ('#'+ url_split.fragment if url_split.fragment else '')))
```
Same here, I think it's better to use stdlib functions for URL construction from parts (see e.g. `urlunsplit` / `urlunparse`)

maramsumanth replied:
I am not be able to do that, because I have ```query_str``` and ```querystr``` to be appended. We can also have the case where ```query_str can be empty```. So, I explicitly defined in the urljoin. Do you have any suggestion @lopuhin ?

lopuhin replied:
> I am not be able to do that, because I have both query_str and querystr to be appended. We can also have the case where query_str can be empty.

@maramsumanth I see, in that case you could still fist concatenate query strings and pass them to urlunsplit, it already has the logic to handle empty parts:

```
In [1]: from urllib.parse import urlunsplit

In [2]: urlunsplit(('a', 'b', 'c', 'd', 'e'))
Out[2]: 'a://b/c?d#e'

In [3]: urlunsplit(('a', 'b', 'c', 'd', ''))
Out[3]: 'a://b/c?d'

In [4]: urlunsplit(('a', 'b', 'c', 'd', ''))
Out[4]: 'a://b/c?d'

In [5]: urlunsplit(('a', 'b', '', '', 'foo'))
Out[5]: 'a://b#foo'
```

maramsumanth replied:
I thought of using urlunsplit, but both urljoin and urlunsplit yield the same thing.

lopuhin commented:
```
http://www.example.com/?a=0&a=2&b=1&c=3#fragment
data = {'a': 1}
```
> Looks like it's wrong request. I tested it by Postman and got 400 status code
> As far as I know all data from GET form should be sent by query string and duplicates of keys can exist because on the server it will be transformed to list. More detail info on SO

@amarynets thanks for review! Could you please clarify what you mean here? I agree that duplicate GET arguments are possible and we should preserve them, but not add new duplicates. Do you mean it's not handled correctly in this PR? If yes, could you please clarify which case is not handled correctly?

@maramsumanth I see that some duplicate cases are checked in tests, but could you please add an explicit test for the case then there is a duplicate GET argument and that it's preserved when we add another unlreated argument? (it seems this is not tested directly).

maramsumanth commented:
@lopuhin Thanks for your review. I have made the changes :) :+1:

lopuhin commented on scrapy/http/request/form.py:
```diff
@@ -33,7 +33,11 @@ def __init__(self, *args, **kwargs):
             else:
-                self._set_url(self.url + ('&' if '?' in self.url else '?') + querystr)
+                url_split = urlsplit(self.url)
+                formdata_key_list = list(dict(parse_qsl(querystr)).keys())
```
Since this is used only for checking filtering in the next line, it would be more efficient with a set instead of a list:
```
formdata_keys = set(dict(parse_qsl(querystr)).keys())
```

maramsumanth commented:
@lopuhin , changes made :+1: , but it is becoming much lengthy, is it ok?

maramsumanth commented:
@whalebot-helmsman @lopuhin. I have made the necessary changes, can you review?

whalebot-helmsman commented:
I think code will be much easier to read if you rename `querystr` to `form_query_str` and `query_str` to `url_query_str`

maramsumanth commented:
Done @whalebot-helmsman :+1:

whalebot-helmsman commented:
@amarynets @lopuhin @maramsumanth I think `FormRequest` behaviour should mirror browser behaviour. To test browser behaviour I create simple HTML file https://gist.githubusercontent.com/whalebot-helmsman/a6c3d2eb78ae57a072af11f8b726ff83/raw/3f281be8ce0f08ea5d81deb443c816b6fbe846f3/forms.html. You can download local copy and test. In case of GET form request both Chrome and Firefox drop `url_query_str` from URL first and add `form_query_str`.

whalebot-helmsman commented:
@maramsumanth Based on browser behaviour I proposed to change implementation in current PR. You up voted this proposal and didn't apply proposed changes. Is there something what prevent you from applying  these changes?

maramsumanth commented:
Thanks for reminding me @whalebot-helmsman .I was working on the Redirect Middleware part of scrapy.
Coming to this PR ,I was little bit confused. In the following example:-
```FormRequest('http://example.com/?id=9', method='GET', formdata={'id': '8'}).url``` should give
```'http://example.com/?id=8'``` i.e it should replace the value if same key is found. I tested in chrome 
   and it gave me the correct output. Can you give me detailed explanation of what is the problem. :)
Thanks

maramsumanth commented:
@whalebot-helmsman , can you please look into this so that I can solve it ASAP. :+1:

whalebot-helmsman commented:
I referenced HTML file as example. It has two forms. First one uses GET request, has `a` and `b` as keys  and targets http://example.com/f?a=1&d=2 URL. After you click `Submit` button you get to http://example.com/f?a=3&b=2. Browser not only replace value for `a` key, but drop value for `d` key. Reformulating in terms of python structures we got
```
FormRequest('http://example.com/?id=9&smth=a', method='GET', formdata={'id': '8'}).url == 'http://example.com/?id=8'
```

maramsumanth commented:
I have Ubuntu OS in which I am using chromium browser.
But for ```FormRequest('http://example.com/?id=9&smth=a', method='GET', formdata={'id': '8'}).url```
my output is ```'http://example.com/?smth=a&id=8'``` which is != ```'http://example.com/?id=8'```

maramsumanth commented:
@whalebot-helmsman , when I tested the HTML file you provided,  I can see that in case of GET form request both Chrome and Firefox drop ```url_query_str``` from URL and add ```form_query_str```. But running through the command line as described in the above comment, I cannot see any problem.

whalebot-helmsman commented:
>I can see that in case of GET form request both Chrome and Firefox drop `url_query_str` from URL and add `form_query_str`. 

Scrapy and related libraries should mirror browser behaviour. 
At this moment in master and in your branch
```
>>> from scrapy.http.request.form import FormRequest;FormRequest('http://example.com/?id=9&smth=a', method='GET', formdata={'id': '8'}).url == 'http://example.com/?id=8'
False
```
This is incorrect representation of browser behaviour. I understand, such behaviour was implemented before your PR. As we started to work on this part we should also fix this.

maramsumanth commented:
Is there any python code to get the default browser the user is having @whalebot-helmsman ?

whalebot-helmsman commented:
> @whalebot-helmsman , when I tested the HTML file you provided, I can see that in case of GET form request both Chrome and Firefox drop `url_query_str` from URL and add `form_query_str`. 

It is a browser behaviour. We agreed on that. 

Scrapy implementation is separate from installed browser. Scrapy behaviour doesn't depend on installed browser. In master branch we preserve both `url_query_str` and `form_query_str`, In branch for this PR we replace values for common keys in `url_query_str` by values from `form_query_str`.
Scrapy's behaviour in both branches doesn't mirror browser behaviour. It is a problem, we should fix it.

I propose to fix it in this PR. 

Current behaviour
```
>>> from scrapy.http.request.form import FormRequest;FormRequest('http://example.com/?id=9&smth=a', method='GET', formdata={'id': '8'}).url
'http://example.com/?smth=a&id=8'
```
Behaviour after fix
```
>>> from scrapy.http.request.form import FormRequest;FormRequest('http://example.com/?id=9&smth=a', method='GET', formdata={'id': '8'}).url
'http://example.com/?id=8'
```

maramsumanth commented:
@whalebot-helmsman Thanks for guiding me.
Till now I was thinking that, we have to somewhere account for the browser name and proceed accordingly. Now I have updated the PR and changed tests as you suggested :)

maramsumanth commented:
> Scrapy behaviour doesn't depend on installed browser

@whalebot-helmsman Thanks for guiding me.
Till now I was thinking that, we have to somewhere account for the browser name and proceed accordingly. Now I have updated the PR and changed tests as you suggested :)

maramsumanth commented:
@whalebot-helmsman , should I change anything or is it safe to merge. :+1: 
Thanks :)

whalebot-helmsman commented:
It is safe to merge now. But merging takes more time and not required by [GSoC rules](http://python-gsoc.org/index.html#apply). We can keep it as it is for now.

Gallaecio commented on tests/test_http_request.py:
```diff
@@ -273,6 +273,18 @@ def assertQueryEqual(self, first, second, msg=None):
+        url = self.request_class('http://www.example.com/?a=0&b=1&c=3#fragment', method='GET', formdata=data).url.split('#')[0]
+        fs = _qs(self.request_class(url, method='GET', formdata=data))
+        self.assertEqual(set(fs[b'a']), {b'one', b'two'})
+        self.assertEqual(fs[b'b'], [b'2'])
```
Maybe also assert that `c` is not in `fs`?

maramsumanth replied:
Also changed test method name :)

maramsumanth commented:
Now its fine :) :+1:
